### ISSUE TITLE
Reduce memory usage when writing using ffmpeg

### ISSUE THREAD

**Issue by djhoese** [imageio/imageio#384]
Thank you for this package. It makes generating movies from python so much easier.

I'm using an imageio writer to generate an mp4 video of some pretty large frames (satellite imagery data). I'm using the `append_data` method to write new frames. Right now after I start adding frames I notice that my memory usage shoots up and I end up going to swap. From what I can tell it seems like ffmpeg isn't consuming the information being written to stdin. Is this something that imageio or ffmpeg can be configured to allow? Is there a way to tell imageio to write to disk and then pass that to ffmpeg? I'm not really an ffmpeg expert so not sure what my options are here or if this can even be avoided easily.

**Comment by almarklein** [comment]
Hi David, this is strange. First thing that pops to mind is that Python generates the data faster than ffmpeg can consume it, filling up the buffer. I don't know from the top of my head whether there is a limit to the buffer, but there probably should be, so that the Python process would block for a bit if ffmpeg can't keep up...

cc @kuchi 

Also see #115, which will hopefully make video a lot easier.

**Comment by djhoese** [comment]
That could be. So theoretically if I added a sleep between my appends then maybe ffmpeg could catch up? I don't see any specific flush in the ffmpeg writer in imageio, but maybe that isn't needed?

**Comment by djhoese** [comment]
Quick update: I tested with a 5 second sleep on a simpler test (less overall frames) and it used a little less memory. Then I tested adding the `-preset slower` to the ffmpeg call and memory usage went way up. Usual computation ~1-1.5GB, with slower preset ~5-6GB in memory.

**Comment by almarklein** [comment]
Could you try setting the `bufsize` argument in the [popen](https://docs.python.org/3.6/library/subprocess.html#subprocess.Popen) call?

**Comment by djhoese** [comment]
Based on the documentation it is set to `io.DEFAULT_BUFFER_SIZE` in python 3.6 (what I'm using) which on my system is 8192. Is there another value that would make sense? But also if buffering is turned on by default shouldn't this mean that python is getting blocked when writing to ffmpeg?

I'm also using dask with 8 workers locally, but my previous tests showed that dask shouldn't be filling up memory. I'll do some more tests today hopefully and see if I can narrow down exactly where things are going. I wonder if I could write to a file on disk and then pass that to ffmpeg as input? Not sure how python/ffmpeg/unix would deal with that buffering-wise.

**Comment by almarklein** [comment]
I think `bufsize` is zero by default. I'd just try a few values to see if there's any change in the memory consumption. Bit of a stab, but if there is a significant change, we can dig deeper.

> I wonder if I could write to a file on disk and then pass that to ffmpeg as input?

I guess you could, but that is not how the plugin works. So it would be more work. Might be easier to wait for the new video plugin.

**Comment by djhoese** [comment]
The documentation you linked to says:

> Changed in version 3.3.1: bufsize now defaults to -1 to enable buffering by default to match the behavior that most code expects. In versions prior to Python 3.2.4 and 3.3.1 it incorrectly defaulted to 0 which was unbuffered and allowed short reads. This was unintentional and did not match the behavior of Python 2 as most code expected.

Still playing with things...

**Comment by djhoese** [comment]
I think I've decided that this isn't imageio's problem. I think this is dask. I tried a few different cases and simplified the images being generated so they are produced as fast as possible. I also tried different presets and while changing the preset might effect the short term memory usage it doesn't affect the overall memory usage of the test script I have. I also tried slicing the large image so dask was computing the whole thing but only 100x100 pixels were being sent to imageio. The result was that the overall/peak memory usage of the script was the same.

I think what is happening and what was my original guess a while back is that imageio/PIPE is blocking the dask writes but the other dask workers are still computing the other frames resulting in a lot of memory usage. There may be some use case to add a to-disk file as input to ffmpeg, but I'll add then when it becomes more of a problem. My workaround right now is to only ask dask to compute 4 frames at a time to reduce the total number of possible frames being held in memory.
